### PR TITLE
Modify Windows OS family types to support platform version in task definitions

### PR DESCRIPTION
**Pull request by alxche2021**
### Summary

This is currently an internal change of generating the correct Windows OS family types, which will be reflected to front end and PV support in task definitions once other changes are made.

### Implementation details

A shared API GetOSFamilyType() is defined in config package to generate Windows OS family types strings. It does not have impact on Linux by keeping its original behavior.

### Testing

Windows binary has been built for the changes. It is being tested on ECS clusters with all supported OS family types. The ECS service is running fine and the log info shows the expected Windows OS family type strings.

New tests cover the changes: No

### Description for the changelog

Feature - Modify Windows OS family types to support platform version in task definitions.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by alxche2021** [comment]
I ll squash my commits into one after all tests are confirmed passing all test cases

File: agent/config/parse_windows.go [hunk 1]
```diff
@@ -113,3 +123,61 @@ var isWindows2016 = func() (bool, error) {
...
+		return "FULL", nil
+	} else if strings.Compare(installationType, InstallationTypeCore) == 0 {
+		return "CORE", nil
+	} else {
+		err := seelog.Errorf("Invalid Installation type")
+		return "", err
+	}
+}
+
+func GetOSFamilyType() string {
```
Comment (mythri-garaga) [inline agent/config/parse_windows.go:138]: Can we add unit test for this?

Comment (alxche2021) [reply agent/config/parse_windows.go:138]: Sure Mythri, I am working on it.

File: agent/config/parse_windows.go [hunk 2]
```diff
@@ -113,3 +123,61 @@ var isWindows2016 = func() (bool, error) {
...
+		return UnknownOSType
+	}
+	releaseId, _, err := key.GetStringValue("ReleaseId")
+	if err != nil {
+		seelog.Errorf("Unable to read registry key, ReleaseId: %v", err)
+		return UnknownOSType
+	}
+
+	if strings.HasPrefix(productName, Windows_Server_2019) {
+		osTypeFormat = fmt.Sprintf(osTypeFormat, "2019", iType)
```
Comment (sparrc) [inline agent/config/parse_windows.go:169]: I find it confusing that you are defining this string as a format string (with `%s`), and then overwriting it as a regular string that ultimately gets returned.

You should create a new variable for the first part of this string similar to how you do for `iType`. In these if-statements, only set the variable to that number. As in: 
```
num := "2019"
```

then you can have just a single return like `return fmt.Sprintf("WINDOWS_SERVER_%s_%s", num, iType)`

File: agent/config/parse_windows.go [hunk 3]
```diff
@@ -113,3 +123,61 @@ var isWindows2016 = func() (bool, error) {
...
+	if err != nil {
+		seelog.Errorf("Unable to read registry key, InstallationType: %v", err)
+		return UnknownOSType
+	}
+	iType, err := getInstallationType(installationType)
+	if err != nil {
+		seelog.Errorf("Invalid Installation type found: %v", err)
+		return UnknownOSType
+	}
+	releaseId, _, err := key.GetStringValue("ReleaseId")
```
Comment (sparrc) [inline agent/config/parse_windows.go:162]: if  i understand correctly, releaseId is only used if this is a windows server datacenter product? can you move getting of this value into that if-statement?

File: agent/config/parse_windows.go [hunk 4]
```diff
@@ -113,3 +123,61 @@ var isWindows2016 = func() (bool, error) {
...
+	} else if strings.Compare(installationType, InstallationTypeCore) == 0 {
+		return "CORE", nil
+	} else {
+		err := seelog.Errorf("Invalid Installation type")
+		return "", err
+	}
+}
+
+func GetOSFamilyType() string {
+	osTypeFormat := "WINDOWS_SERVER_%s_%s"
```
Comment (sparrc) [inline agent/config/parse_windows.go:139]: define this either as a constant in the whole package or not at all

**Review by sparrc — CHANGES_REQUESTED** [review]

File: agent/api/ecsclient/client.go [hunk 5]
```diff
@@ -325,10 +325,12 @@ func validateRegisteredAttributes(expectedAttributes, actualAttributes []*ecs.At
 }
 
 func (client *APIECSClient) getAdditionalAttributes() []*ecs.Attribute {
+	osFamilyStr := config.GetOSFamilyType()
+	seelog.Infof("Server OSFamily string: %s", osFamilyStr)
```
Comment (sparrc) [inline agent/api/ecsclient/client.go:329]: this seems like more of a Debugf statement

File: agent/config/parse_windows.go [hunk 6]
```diff
@@ -16,22 +16,32 @@
...
 const (
 	// envSkipDomainJoinCheck is an environment setting that can be used to skip
 	// domain join check validation. This is useful for integration and
 	// functional-tests but should not be set for any non-test use-case.
 	envSkipDomainJoinCheck = "ZZZ_SKIP_DOMAIN_JOIN_CHECK_NOT_SUPPORTED_IN_PRODUCTION"
+
+	// The following constants are used to generate Windows OS family types
+	ReleaseId2004SAC          = "2004"
+	ReleaseId1909SAC          = "1909"
+	Windows_Server_2019       = "Windows Server 2019"
```
Comment (sparrc) [inline agent/config/parse_windows.go:39]: can you remove the underscores and change these to WindowsServer2019 for consistency?

File: agent/config/parse_windows.go (hunk 1 above)
Comment (sparrc) [inline agent/config/parse_windows.go:138]: It seems if any step of this GetOSFamilyType function fails it seems we will just return `unknown`, should we maybe return `windows` (ie, `config.OSType`) to keep the behavior the same as before?

**Comment by sparrc** [comment]
closing as this change was implemented in https://github.com/aws/amazon-ecs-agent/pull/2859

**Pull request closed without merging**